Vloginov/im2latex training (#339)
openvinotoolkit/training_extensions: pytorch_toolkit/formula_recognition/README.md

- `vocab_path` - path where vocab file is stored
- `train_transforms_list` and
- `val_transforms_list` - here you can describe set of desirable transformations for train and validation datasets respectively
- `epochs` - number of epochs to train
- `clip_grad` - maximum possible value for gradient
- `old_model` - use this flag if you want to traing model trained in the previous versions of this framework.

One can point to pre-trained model [checkpoint](https://download.01.org/opencv/openvino_training_extensions/models/text_spotter/model_step_200000.pth) inside configuration file to start training from pre-trained weights. Change `configs/train_config.yml`:
```
...
model_path: <path_to_weights>
...
```

If the model was marked `old_model`, that means the model was trained in older version of this framework (concretely, model checkpoint keys are different from keys used in model now), so if you want to use this model in any context, point out this fact in `config`:
```
...
old_model: true
...
```

## Evaluation

`tools/test.py` script is designed for quality evaluation of im2latex models.

### PyTorch

Config file of the evaluation is similiar to train config:

```bash

> python tools/test.py --config configs/eval_config.yml

Reviewer: Where should we write the path to the checkpoint? to the `model_path` field? or in another field? could you write it?

Author: The names of the parameters in the evaluation config are the same as in train one, so there is no need to duplicate them. Anyway, all the four configs will be merged into one file.

```
Evaluation process is the following:
1. Run the model and get predictions
2. Render predictions from the first step into images of the formulas
3. Compare images.
The third step is very important because im LaTeX language one can write different formulas that are looking the same. Example:

> `s^{12]_{i}` and `s_{i]^{12}` looking the same: $s^{i}_{12}$ vs. $s_{12}^{i}$

Reviewer: There is a bug in the formulas here: `]` instead of `}`

Reviewer: ```suggestion
`s^{12}_{i}` and `s_{i}^{12}` looking the same: both of them are rendered as ![equation](https://latex.codecogs.com/gif.latex?%5Cbg_white%20s%5E%7Bi%7D_%7B12%7D)
```

# PyTorch realization of the Im2Markup

This repository contains inference and training code for Im2LaTeX models.

> Source [repository](https://github.com/harvardnlp/im2markup/). This repository is a fork of [PyTorch realization](https://github.com/luopeixiang/im2latex/)

Reviewer 2: It is not a fork

> Models code is designed to enable ONNX\* export and inference on CPU via OpenVINO™.

Reviewer 2: CPU only?

Author: fixed

## Setup

### Prerequisites

* Ubuntu\* 16.04
* Python\* 3.7 or newer
* PyTorch\* (1.4.0)
* CUDA\* 10.1

> * OpenVINO™ 2020.1 with Python API

Reviewer 2: too old already

Author: What exactly is too old?

Reviewer 3: OpenVINO version here is outdated. Please put a modern one (I hope it's compatible with the code).

Author: Done

### Installation

Create and activate virtual environment:

```bash
bash init_venv.sh
```

### Download Datasets

> For training model One has to have dataset. Dataset format is similiar to [im2latex-100k](https://zenodo.org/record/56198#.X2NDQ2gzaUl). Main structure of the dataset is following:

Reviewer 2: one

> For training model One has to have dataset. Dataset format is similiar to [im2latex-100k](https://zenodo.org/record/56198#.X2NDQ2gzaUl). Main structure of the dataset is following:

Reviewer 2: similar

* `formulas_file` - file with one formula per line
* `images_folder` - folder containing input images

> * `split_file` - the most important part of the dataset. This file contains `file_name` (tab symbol) `formula_idx` per line connecting corresponding index of the formula in the file with formulas and particular image with `image_name`. Example:

Reviewer 2: Is 'the most important part of the dataset.' sentence needed?

Author: Reformulated

    ```
    11.png  11
    34.png  34
    ```
    There should be at least two such files: `train_filter.lst` and `validate_filter.lst`

> **NOTE**:

> > By default code asserts to see dataset structure with the following names:

Reviewer 2: is `asserts` a proper verb?

Author: Reformulated

> `images_processed` - folder with images
> `formulas.norm.lst` - file with preprocessed formulas, for details, refer to [im2markup](https://github.com/harvardnlp/im2markup) repository
> `validate_filter.lst` and `train_filter.lst` - corresponding splits of the data.

## Training

> To train Text Spotter model run:

Reviewer 2: Text Spotter?

Author: Fixed

```bash
python3 tools/train.py --config configs/train_config.yml --work_dir <path to work dir>
```
Work dir is used to store information about learning: saved model checkpoints, logs.

> ### Description of the possible option in train config:

Reviewer 2: of possible options in training configuration

Author: Fixed

 - `backbone_config`:
    * `arch`: type of the architecture (if backbone_type is resnet). For mor details, please, refer to [ResnetLikeBackBone](im2latex/models/backbones/resnet.py)

>     * `disable_layer_3` and

Reviewer 2: and ?

Author: reformulated

>     * `disable_layer_4` - disables layer 3 and 4 in resnetlike backbone

Reviewer 2: resnet-like

Author: fixed

    * `enable_last_conv` - enables additional convolution layer to adjust number of output channels to the number of input channels in the LSTM
    * `in_lstm_ch` - number of input LSTM channels, used for `last_conv`

> - `backbone_type`: `resnet` for resnetlike backbone or anything else for original backbone from [im2markup](https://arxiv.org/pdf/1609.04938.pdf) paper

Reviewer 2: ditto

Author: fixed

- `batch_size` - batch size used for training
- `device` - device for training, used in pytorch .to() method. Possible options: 'cuda', 'cpu', etc. `cpu` is used by default.
- `head` - configuration of the text recognition head
    * `beam_width` - witdth used in beam search. 0 - do not use beam search, 1 and more - use beam search with corresponding number of possible tracks.

>     * `dec_rnn_h` - number of channels on decoding

Reviewer 2: in?

Author: fixed

    * `emb_size` - dimension of the embedding
    * `enc_rnn_h` - number of channels in encoding

>     * `in_lstm_ch` - number of input in lstm channels, should be equal to `backbone_config.in_lstm_ch`

Reviewer 2: wat

Author: renamed parameter

>     * `max_len` - maximum possible length of the formula

Reviewer 2: which formula (the?)

Author: reformulated

    * `n_layer` - describe
- `learning_rate` - learining rate
- `log_path` - path to store training logs
- `model_path` - path for model (if one wants to aftertune model)
- `optimizer` - Adam or SGD
- `save_dir` - dir to save checkpoints

> - `train_paths` - list of paths from where get training data (if more than one path is specified, datasets are concatenated). If one wants to concatenate more than one instance of the desirable dataset, this dataset shoulb be set several times.

Reviewer 2: `this dataset shoulb be set several times.` what?

Author: fixed

- `val_path` - path for validation data
- `vocab_path` - path where vocab file is stored

> - `train_transforms_list` and

Reviewer 2: ?

Author: ?

Author: reformulated

- `val_transforms_list` - here you can describe set of desirable transformations for train and validation datasets respectively
- `epochs` - number of epochs to train
- `clip_grad` - maximum possible value for gradient

> - `old_model` - use this flag if you want to traing model trained in the previous versions of this framework.

Reviewer 2: ?

Author: ?

Author: removed

One can point to pre-trained model [checkpoint](https://download.01.org/opencv/openvino_training_extensions/models/text_spotter/model_step_200000.pth) inside configuration file to start training from pre-trained weights. Change `configs/train_config.yml`:
```
...
model_path: <path_to_weights>
...
```

> If the model was marked `old_model`, that means that means that model was trained in older version of this framework (concretly, model checkpoint keys are different from keys used in model now), so if you want to use this model in any context, point out this fact in `config`:

Reviewer 2: `that means that means`

Author: fixed

> If the model was marked `old_model`, that means that means that model was trained in older version of this framework (concretly, model checkpoint keys are different from keys used in model now), so if you want to use this model in any context, point out this fact in `config`:

Reviewer 2: concretly

Author: fixed

```
...
old_model: true
...
```

## Evaluation

`tools/test.py` script is designed for quality evaluation of im2latex models.

### PyTorch

Config file of the evaluation is similiar to train config:

```bash
python tools/test.py --config configs/eval_config.yml
```
Evaluation process is the following:
1. Run the model and get predictions
2. Render predictions from the first step into images of the formulas
3. Compare images.
The third step is very important because im LaTeX language one can write different formulas that are looking the same. Example:
`s^{12]_{i}` and `s_{i]^{12}` looking the same: $s^{i}_{12}$ vs. $s_{12}^{i}$
That is why we cannot just compare text predictions one-by-one, we have to render images and compare them.

## Demo

In order to see how trained model works using OpenVINO™ please refer to [Formula recognition Python* Demo](https://github.com/opencv/open_model_zoo/tree/develop/demos/python_demos/formula_recognition_demo/). Before running the demo you have to export trained model to IR. Please see below how to do that.

> If you want to se how trained pytorch model is working, you can run `tools/demo.py` script with correct `config` file. Fill in the `input_images` variable with the paths to desired images. For every image in this list, model will predict the formula and print it into the terminal.

Reviewer 2: see

Author: fixed

> If you want to se how trained pytorch model is working, you can run `tools/demo.py` script with correct `config` file. Fill in the `input_images` variable with the paths to desired images. For every image in this list, model will predict the formula and print it into the terminal.

Reviewer 2: probably wrong usage of `fill in <> with`

# PyTorch realization of the Im2Markup

This repository contains inference and training code for Im2LaTeX models.

> Source [repository](https://github.com/harvardnlp/im2markup/). This repository is based on a [PyTorch realization](https://github.com/luopeixiang/im2latex/)

Author: Rephrased

Models code is designed to enable ONNX\* export and inference on CPU\GPU via OpenVINO™.

## Setup

### Prerequisites

* Ubuntu\* 16.04
* Python\* 3.7 or newer
* PyTorch\* (1.4.0)

> * CUDA\* 10.1

Author: Removed cuda from requirements

* OpenVINO™ 2020.1 with Python API

### Installation

Create and activate virtual environment:

```bash
bash init_venv.sh
```

### Download Datasets

For training model one has to have dataset. Dataset format is similiar to [im2latex-100k](https://zenodo.org/record/56198#.X2NDQ2gzaUl). Main structure of the dataset is following:
* `formulas_file` - file with one formula per line
* `images_folder` - folder containing input images

> * `split_file` - this file contains `file_name` (tab symbol) `formula_idx` per line connecting corresponding index of the formula in the file with formulas and particular image with `image_name`. Example:

Author: Fixed

    ```
    11.png  11
    34.png  34
    ```
    There should be at least two such files: `train_filter.lst` and `validate_filter.lst`

> **NOTE**:
> By default the following structure of the dataset is assumed:
> `images_processed` - folder with images
> `formulas.norm.lst` - file with preprocessed formulas, for details, refer to [im2markup](https://github.com/harvardnlp/im2markup) repository
> `validate_filter.lst` and `train_filter.lst` - corresponding splits of the data.

## Training

To train formula-recognition model run:

```bash
python3 tools/train.py --config configs/config.yml --work_dir <path to work dir>
```
Work dir is used to store information about learning: saved model checkpoints, logs.

### Description of possible options in config:
The config file is divided into 4 sections: train, eval, export, demo. In every section there could exist common parameter as well as unique parameters.
#### Common parameters:

>  - `backbone_config`:

Author: It is planned to make configs and interface like in model template in the future versions.

>  - `backbone_config`:

Reviewer 3: Good to know.

- `input_images` - list of paths for input images

#### Export-specific parameters
These are parameters used for model export to ONNX & OpenVINO™ IR:
- `res_encoder_name` - filename to save the converted encoder model (with `.onnx` postfix)
- `res_decoder_name` - filename to save the converted decoder model (with `.onnx` postfix)
- `dummy_input` - input image used for tracing PT model and creating onnx graph. This image is also used to compare that outputs from ONNX model and PyTorch model are the same.
- `input_shape_decoder` - list of dimensions describing input shape for encoder for OpenVINO IR conversion.
- `export_ir` - Set this flag to `true` to export model to the OpenVINO IR. For details refer to [convert to IR section](#convert-to-ir)
- `verbose_export` - Set this flag to `true` to perform verbose export (i.e. print model optimizer commands to terminal)

One can point to pre-trained model [checkpoint](https://download.01.org/opencv/openvino_training_extensions/models/text_spotter/model_step_200000.pth) inside configuration file to start training from pre-trained weights. Change `configs/config.yml`:
```
...
model_path: <path_to_weights>
...
```

> If the model was marked `old_model`, that means the model was trained in older version of this framework (concretely, model checkpoint keys are different from keys used in model now), so if you want to use this model in any context, point out this fact in `config`:

Author: Removed `old_format` parameter, add handling of the old model in the `load_weights()`

```
...
old_model: true
...
```

## Evaluation

`tools/test.py` script is designed for quality evaluation of im2latex models.

### PyTorch

Config file of the evaluation is similiar to train config:

```bash

> python tools/test.py --config configs/config.yml

Author: Fixed

    * `disable_layer_3` and `disable_layer_4` - disables layer 3 and 4 in resnet-like backbone
    * `enable_last_conv` - enables additional convolution layer to adjust number of output channels to the number of input channels in the LSTM. Optional. Default: false.
    * `in_lstm_ch` - number of input LSTM channels, used only for `last_conv`.
- `backbone_type`: `resnet` for resnet-like backbone or anything else for original backbone from [im2markup](https://arxiv.org/pdf/1609.04938.pdf) paper. Optional. Default is `resnet`
- `head` - configuration of the text recognition head. All of the following parameters have default values, you can check them in [text reconition head](im2latex/models/text_recognition_heads/attention_based.py)
    * `beam_width` - witdth used in beam search. 0 - do not use beam search, 1 and more - use beam search with corresponding number of possible tracks.
    * `dec_rnn_h` - number of channels in decoding
    * `emb_size` - dimension of the embedding
    * `encoder_hidden_size ` - number of channels in encoding

>     * `encoder_input_size ` - number of input in lstm channels, should be equal to `backbone_config.in_lstm_ch`

Author: Paraphrased, thanks

    * `max_len` - maximum possible length of the predicted formula
    * `n_layer` - number of layers in the trainable initial hidden state for each row

> - `model_path` - path for model

Author: fixed

> - `val_path` - path for validation data

Author: fixed

    vocab_path: vocabs/vocab_handwritten_polynomials.json
    ```
The model can be used for recognizing handwritten polynomial equations.
All the above models can be used for aftertuning or as ready for inference models. To provide maximum quality at recognizing formulas, it is highly recommended to preprocess image - simply binarize it, you can find corresponding prepocessing at [this file](im2latex/data/utils.py). Just state the desired preprocessing at the corresponding section of the config file (train, eval, etc).

#### Evaluation-specific parameters
- `split_file` - name of the file with labels (note: physical file name should end with `_filter.lst`). Default is `validate`
- `target_metric` - target value of the metric. Used in tests. For test to pass, result value should be greater or equal than `target_metric`

#### Demo-specific parameters
- `input_images` - list of paths for input images

#### Export-specific parameters
These parameters are used for model export to ONNX & OpenVINO™ IR:
- `res_encoder_name` - filename to save the converted encoder model (with `.onnx` postfix)
- `res_decoder_name` - filename to save the converted decoder model (with `.onnx` postfix)

> - `dummy_input` - input image used for tracing PT model and creating onnx graph. This image is also used to compare that outputs from ONNX model and PyTorch model are the same.

Author: removed this argument

> # PyTorch realization of the Im2Markup

Reviewer 2: I think `# Formula recognition` is better to use

> This repository is based on a [PyTorch realization](https://github.com/luopeixiang/im2latex/) of the code from the original [repository](https://github.com/harvardnlp/im2markup/).

Reviewer 2: This `code` but not `repo` I think

Models code is designed to enable ONNX\* export and inference on CPU\GPU via OpenVINO™.

## Setup

### Prerequisites

* Ubuntu\* 16.04
* Python\* 3.7 or newer
* PyTorch\* (1.4.0)
* OpenVINO™ 2020.4 with Python API

### Installation

Create and activate virtual environment:

```bash
bash init_venv.sh
```

### Download Datasets

> For training model one has to have dataset. Dataset format is similiar to [im2latex-100k](https://zenodo.org/record/56198#.X2NDQ2gzaUl). Main structure of the dataset is following:

Reviewer 2: `For training model one has to have dataset. ` looks obvious and redundant

    * `enable_last_conv` - enables additional convolution layer to adjust number of output channels to the number of input channels in the LSTM. Optional. Default: false.
    * `in_lstm_ch` - number of input LSTM channels, used only for `last_conv`.
- `backbone_type`: `resnet` for resnet-like backbone or anything else for original backbone from [im2markup](https://arxiv.org/pdf/1609.04938.pdf) paper. Optional. Default is `resnet`
- `head` - configuration of the text recognition head. All of the following parameters have default values, you can check them in [text reconition head](im2latex/models/text_recognition_heads/attention_based.py)
    * `beam_width` - witdth used in beam search. 0 - do not use beam search, 1 and more - use beam search with corresponding number of possible tracks.
    * `dec_rnn_h` - number of channels in decoding
    * `emb_size` - dimension of the embedding
    * `encoder_hidden_size ` - number of channels in encoding
    * `encoder_input_size ` - number of input in lstm channels, should be equal to `backbone_config.in_lstm_ch`
    * `max_len` - maximum possible length of the predicted formula

>     * `n_layer` - describe

Reviewer 2: what does it describe?

Author: Updated readme, it was a mark for me to describe this parameter

- `model_path` - path for model
- `val_path` - path for validation data
- `vocab_path` - path where vocab file is stored

> - `val_transforms_list` - here you can describe set of desirable transformations for validation datasets respectively An example is given in the config file, for other options, please, refer to [constructor of transforms (section `create_list_of_transforms`)](im2latex/data/utils.py)

Reviewer 2: respectively.

> - `device` - device for training, used in PyTorch .to() method. Possible options: 'cuda', 'cpu', etc. `cpu` is used by default.

Reviewer 2: which cases can `etc` cover here?

Author: Removed `etc`

#### Training-specific parameters
In addition to common parameters you can specify the following arguments:
- `batch_size` - batch size used for training
- `learning_rate` - learining rate
- `log_path` - path to store training logs
- `optimizer` - Adam or SGD
- `save_dir` - dir to save checkpoints

> - `train_paths` - list of paths from where get training data (if more than one path is specified, datasets are concatenated). If one wants to concatenate more than one instance of the desirable dataset, this dataset should be specified several times.

Reviewer 2: to get

> - `train_transforms_list` - similiar to `val_transforms_list`

Reviewer 2: similar

    * digits, letters, upper indices are supported
    * to use this model, please, change these field in the config file:
    ```
    backbone_config:
        disable_layer_3: false
    model_path: <path to the model>
    vocab_path: vocabs/vocab_handwritten_polynomials.json
    ```
The model can be used for recognizing handwritten polynomial equations.
All the above models can be used for aftertuning or as ready for inference models. To provide maximum quality at recognizing formulas, it is highly recommended to preprocess image - simply binarize it, you can find corresponding prepocessing at [this file](im2latex/data/utils.py). Just state the desired preprocessing at the corresponding section of the config file (train, eval, etc).

#### Evaluation-specific parameters
- `split_file` - name of the file with labels (note: physical file name should end with `_filter.lst`). Default is `validate`
- `target_metric` - target value of the metric. Used in tests. For test to pass, result value should be greater or equal than `target_metric`

#### Demo-specific parameters
- `input_images` - list of paths for input images

#### Export-specific parameters

> These are parameters used for model export to ONNX & OpenVINO™ IR:

Reviewer 2: these parameters are used

- `res_encoder_name` - filename to save the converted encoder model (with `.onnx` postfix)
- `res_decoder_name` - filename to save the converted decoder model (with `.onnx` postfix)
- `dummy_input` - input image used for tracing PT model and creating onnx graph. This image is also used to compare that outputs from ONNX model and PyTorch model are the same.
- `input_shape_decoder` - list of dimensions describing input shape for encoder for OpenVINO IR conversion.
- `export_ir` - Set this flag to `true` to export model to the OpenVINO IR. For details refer to [convert to IR section](#convert-to-ir)
- `verbose_export` - Set this flag to `true` to perform verbose export (i.e. print model optimizer commands to terminal)

> 

Reviewer 2: redundant empty lines

## Evaluation

`tools/test.py` script is designed for quality evaluation of im2latex models.

### PyTorch

> Config file of the evaluation is similiar to train config:

Reviewer 2: similar

# PyTorch realization of the Formula Recognition

This code is based on a [PyTorch realization](https://github.com/luopeixiang/im2latex/) of the code from the original [repository](https://github.com/harvardnlp/im2markup/).
Models code is designed to enable ONNX\* export and inference on CPU\GPU via OpenVINO™.

## Setup

### Prerequisites

* Ubuntu\* 16.04
* Python\* 3.7 or newer
* PyTorch\* (1.4.0)
* OpenVINO™ 2020.4 with Python API

### Optional prerequisites

#### Install required packages for evaluation
These packages are used for rendering images while evaluation and demo.

```bash

> bash sudo apt-get update &&

Reviewer 3: ```suggestion
sudo apt-get update &&
```

Author: Fixed

  apt-get install -y --no-install-recommends \
    texlive \
    imagemagick \
    ghostscript
```

### Installation

Create and activate virtual environment:

```bash
bash init_venv.sh
```

> ### Download Datasets

Author: Added more information about creating custom dataset

Dataset format is similar to [im2latex-100k](https://zenodo.org/record/56198#.X2NDQ2gzaUl). Main structure of the dataset is following:
* `formulas_file` - file with one formula per line
* `images_folder` - folder containing input images
* `split_file` - this file contains `image_name` (tab symbol) `formula_idx` per line connecting corresponding index of the formula in the file with formulas and particular image with `image_name`. Example:
    ```
    11.png  11
    34.png  34
    ...
    ```
    There should be at least two such files: `train_filter.lst` and `validate_filter.lst`

> **NOTE**:
> By default the following structure of the dataset is assumed:
> `images_processed` - folder with images

> > `formulas.norm.lst` - file with preprocessed formulas, for details, refer to [im2markup](https://github.com/harvardnlp/im2markup) repository

Author: Changed link to preprocessing script

> `validate_filter.lst` and `train_filter.lst` - corresponding splits of the data.

> 

Author: Added more information about creating custom dataset

## Training

To train formula recognition model run:

```bash
python tools/train.py --config configs/medium_config.yml --work_dir <path to work dir>
```
Work dir is used to store information about learning: saved model checkpoints, logs.

### Description of possible options in config:

> The config file is divided into 5 sections: common, train, eval, export, demo. Common parameters (like path to the model) are stored, respectively, in common section. Unique parameters (like learning rate) are stored in other specific sections.

Author: Added information about checking config keys

#### Common parameters:

>  - `backbone_config`:

Author: Fixed

    * `arch`: type of the architecture (if backbone_type is resnet). For more details, please, refer to [ResnetLikeBackBone](im2latex/models/backbones/resnet.py)

>     * `disable_layer_3` and `disable_layer_4` - disables layer 3 and 4 in resnet-like backbone

Reviewer: Could you add a short description what is it?

At least, I'd propose to add info something like that     
`(note that ResNet backbone from torchvision consists of 4 block of layers, each of them increase the number of channels and decrease the spatial dimensionality, these parameters allow to swithch off the last of them)`

Author: Added information about layer_3 & layer_4

    * `enable_last_conv` - enables additional convolution layer to adjust number of output channels to the number of input channels in the LSTM. Optional. Default: false.

>     * `output_channels` - number of output channels channels. If `last_conv` is enabled, this parameter should be equal to `head.encoder_input_size`, otherwise it should be equal to actual number of output channels of the backbone.

Reviewer: ```suggestion
    * `output_channels` - number of output channels channels. If `lenable_ast_conv` is set, this parameter should be equal to `head.encoder_input_size`, otherwise it should be equal to actual number of output channels of the backbone.
```

Author: Fixed

- `backbone_type`: `resnet` for resnet-like backbone or anything else for original backbone from [im2markup](https://arxiv.org/pdf/1609.04938.pdf) paper. Optional. Default is `resnet`
- `head` - configuration of the text recognition head. All of the following parameters have default values, you can check them in [text reconition head](im2latex/models/text_recognition_heads/attention_based.py)

>     * `beam_width` - witdth used in beam search. 0 - do not use beam search, 1 and more - use beam search with corresponding number of possible tracks.

Reviewer: ```suggestion
    * `beam_width` - width used in beam search. 0 - do not use beam search, 1 and more - use beam search with corresponding number of possible tracks.
```

Author: Fixed

    * `dec_rnn_h` - number of channels in decoding
    * `emb_size` - dimension of the embedding
    * `encoder_hidden_size ` - number of channels in encoding
    * `encoder_input_size ` - number of channels in the lstm input, should be equal to `backbone_config.output_channels`
    * `max_len` - maximum possible length of the predicted formula
    * `n_layer` - number of layers in the trainable initial hidden state for each row

> - `model_path` - path to the model

Reviewer: Please, add couple of words here, or during the first reading it is not clear    
(model to store? to read? model in ONNX or another format?)

Author: Added more information about model checkpoint

> - `val_path` - path to the validation data

Author: Done

> - `vocab_path` - path where vocab file is stored

Author: Added info about what is vocab file

- `train_transforms_list` - similar to `val_transforms_list`
- `epochs` - number of epochs to train

One can use some pretrained models. Right now two models are available:
* medium model:
    * [checkpoint link](https://download.01.org/opencv/openvino_training_extensions/models/formula_recognition/medium_photograped_0185.pth)
    * digits, letters, some greek letters, fractions, trigonometric operations are supported; for more details, please, look at corresponding vocab file
    * to use this model, just set the correct value to the `model_path` field in the corresponding config file:
    ```
    model_path: <path to the model>
    ```
The model can be used for recognizing both rendered and scanned formulas (e.g. from a scanner or from a phone camera)

* handwritten polynomials model:
    * [checkpoint](https://download.01.org/opencv/openvino_training_extensions/models/formula_recognition/polynomials_handwritten_0166.pth)
    * digits, letters, upper indices are supported
    * to use this model, please, change model path in the corresponding config file:
    ```
    model_path: <path to the model>
    ```
The model can be used for recognizing handwritten polynomial equations.

> All the above models can be used for aftertuning or as ready for inference models. To provide maximum quality at recognizing formulas, it is highly recommended to preprocess image - simply binarize it, you can find corresponding prepocessing at [this file](im2latex/data/utils.py). Sample images in the [data](../../data) section of this repo are already preprocessed, you can look at the examples. If you want to use our own dataset, just state the desired preprocessing at the corresponding section of the config file (train, eval, etc).

Reviewer: The words `find corresponding prepocessing at...` are a bit confusing -- there are a lot of preprocessing functions in that file, which function do you mean? Could you refer to a concrete function?

And, please, split the line 🙂

Author: Paraphrased this paragraph

- `export_ir` - Set this flag to `true` to export model to the OpenVINO IR. For details refer to [convert to IR section](#convert-to-ir)
- `verbose_export` - Set this flag to `true` to perform verbose export (i.e. print model optimizer commands to terminal)

## Evaluation

`tools/test.py` script is designed for quality evaluation of formula-recognition models.

### PyTorch

For example, one can run evaluation process using config for `medium` model.
```bash
python tools/test.py --config configs/medium_config.yml
```
Evaluation process is the following:
1. Run the model and get predictions
2. Render predictions from the first step into images of the formulas
3. Compare images.
The third step is very important because in LaTeX language one can write different formulas that are looking the same. Example:
`s^{12}_{i}` and `s_{i}^{12}` looking the same: both of them are rendered as ![equation](https://latex.codecogs.com/gif.latex?%5Cbg_white%20s%5E%7Bi%7D_%7B12%7D)
That is why we cannot just compare text predictions one-by-one, we have to render images and compare them.

## Demo

> In order to see how trained model works using OpenVINO™ please refer to [Formula recognition Python* Demo](https://github.com/opencv/open_model_zoo/tree/develop/demos/python_demos/formula_recognition_demo/). Before running the demo you have to export trained model to IR. Please, see below how to do that.

Reviewer: ```suggestion
In order to see how trained model works using OpenVINO™ please refer to [Formula recognition Python\* Demo](https://github.com/opencv/open_model_zoo/tree/develop/demos/python_demos/formula_recognition_demo/). Before running the demo you have to export trained model to IR. Please, see below how to do that.
```

Author: Fixed

# PyTorch realization of the Formula Recognition

This code is based on a [PyTorch realization](https://github.com/luopeixiang/im2latex/) of the code from the original [repository](https://github.com/harvardnlp/im2markup/).
Models code is designed to enable ONNX\* export and inference on CPU\GPU via OpenVINO™.

## Setup

### Prerequisites

* Ubuntu\* 16.04
* Python\* 3.7 or newer
* PyTorch\* (1.4.0)
* OpenVINO™ 2020.4 with Python API

### Optional prerequisites

#### Install required packages for evaluation
These packages are used for rendering images while evaluation and demo.

```bash
sudo apt-get update &&
  apt-get install -y --no-install-recommends \
    texlive \
    imagemagick \
    ghostscript
```

### Installation

Create and activate virtual environment:

```bash
bash init_venv.sh
```

### Download or Prepare Datasets

> 

Reviewer: Vladimir, if we add a new subsection `#### Vocabulary files` as I wrote in my previous comment, then we should add a new subsection here (otherwise the tree of the document won't be "balanced").

My suggestion is to add a new subsection header at this point as follows:
```
#### Dataset format
```

Dataset format is similar to [im2latex-100k](https://zenodo.org/record/56198#.X2NDQ2gzaUl). Main structure of the dataset is following:

> * `formulas_file` - file with one formula per line

Reviewer: Vladimir, I'd move details on preprocessing here
```suggestion
* `formulas_file` - file with one formula per line.
    If you want to use your own dataset, formulas should be preprocessed.
    For details, refer to [this script](https://github.com/harvardnlp/im2markup/blob/master/scripts/preprocessing/preprocess_formulas.py).
```

* `images_folder` - folder containing input images
* `split_file` - this file contains `image_name` (tab symbol) `formula_idx` per line connecting corresponding index of the formula in the file with formulas and particular image with `image_name`. Example:
    ```
    11.png  11
    34.png  34
    ...
    ```
    There should be at least two such files: `train_filter.lst` and `validate_filter.lst`

> You can prepare your own dataset in the same format as above. Samples of the dataset can be found [here](../../data/formula_recognition). When you prepare your own dataset with `formulas.norm.lst` file, you will have to create a vocabulary file for this dataset. Vocabulary file is a special file which is used to cast token ids to human readable tokens and vice versa. Like letters and digits in the natural language, tokens here are atomic units of the latex language (e.g. `\\sin`, `1`, `\\sqrt`, etc). You can find an example in the [vocabs folder](./vocabs/) of this project. Use [this script](./tools/make_vocab.py) to create vocab file from your own formulas file. The script will read the formulas and create the vocabulary from the formulas used in train split of the dataset.

Reviewer: Vladimir, IMHO vocab is sufficiently important part of interface of the tools. 
At the moment description what is vocab is hidden a bit -- it is inside a non-top paragraph, and the sentence what is vocab started not in beginning of the paragraph.

So, I'd propose to make a separate small subsection about vocab

```suggestion
You can prepare your own dataset in the same format as above. Samples of the dataset can be found [here](../../data/formula_recognition). 

#### Vocabulary files

When you prepare your own dataset with `formulas.norm.lst` file, you will have to create a vocabulary file for this dataset.
Vocabulary file is a special file which is used to cast token ids to human readable tokens and vice versa. Like letters and digits in the natural language, tokens here are atomic units of the latex language (e.g. `\\sin`, `1`, `\\sqrt`, etc). You can find an example in the [vocabs folder](./vocabs/) of this project. 
Use [this script](./tools/make_vocab.py) to create vocab file from your own formulas file. The script will read the formulas and create the vocabulary from the formulas used in train split of the dataset.
```

P.S. Vladimir, I split one line into several lines -- it is much more convenient for review, whereas github will merge them into one line (it is a standard Markdown feature).

> **NOTE**:
> By default the following structure of the dataset is assumed:
> `images_processed` - folder with images
> `formulas.norm.lst` - file with preprocessed formulas. If you want to use your own dataset, formulas should be preprocessed. For details, refer to [this script](https://github.com/harvardnlp/im2markup/blob/master/scripts/preprocessing/preprocess_formulas.py).

> > `validate_filter.lst` and `train_filter.lst` - corresponding splits of the data.

Reviewer: I'd propose to move this whole note above, just before the subsection `#### Vocabulary file`

## Training

To train formula recognition model run:

```bash
python tools/train.py --config configs/medium_config.yml --work_dir <path to work dir>
```
Work dir is used to store information about learning: saved model checkpoints, logs.

### Description of possible options in config:
The config file is divided into 5 sections: common, train, eval, export, demo. Common parameters (like path to the model) are stored, respectively, in common section. Unique parameters (like learning rate) are stored in other specific sections. Unique parameters and common parameters are mutually exclusive.
#### Common parameters:
- `backbone_config`:
    * `arch`: type of the architecture (if backbone_type is resnet). For more details, please, refer to [ResnetLikeBackBone](im2latex/models/backbones/resnet.py)

>     * `disable_layer_3` and `disable_layer_4` - disables layer 3 and 4 in resnet-like backbone. ResNet backbone from the torchvision module consists of 4 block of layers, each of them increase the number of channels and decrease the spatial dimensionality. These parameters allow to switch off the 3rd and the 4th of such layer, respectively.

Reviewer: ```suggestion
    * `disable_layer_3` and `disable_layer_4` - disables layer 3 and 4 in resnet-like backbone. ResNet backbone from the torchvision module consists of 4 block of layers, each of them increase the number of channels and decrease the spatial dimensionality. These parameters allow to switch off the 3rd and the 4th of such layers, respectively.
```

- `device` - device for training, used in PyTorch .to() method. Possible options: 'cuda', 'cpu'. `cpu` is used by default.
#### Training-specific parameters
In addition to common parameters you can specify the following arguments:
- `batch_size` - batch size used for training
- `learning_rate` - learining rate
- `log_path` - path to store training logs
- `optimizer` - Adam or SGD
- `save_dir` - dir to save checkpoints
- `train_paths` - list of paths from where to get training data (if more than one path is specified, datasets are concatenated). If one wants to concatenate more than one instance of the desirable dataset, this dataset should be specified several times.
- `val_path` - path to the validation data
- `train_transforms_list` - similar to `val_transforms_list`
- `epochs` - number of epochs to train

One can use some pretrained models. Right now two models are available:
* medium model:
    * [checkpoint link](https://download.01.org/opencv/openvino_training_extensions/models/formula_recognition/medium_photograped_0185.pth)

>     * digits, letters, some greek letters, fractions, trigonometric operations are supported; for more details, please, look at corresponding vocab file

Reviewer: Please, make `corresponding vocab file` a link to the file in the repo

    ```
The model can be used for recognizing both rendered and scanned formulas (e.g. from a scanner or from a phone camera)

* handwritten polynomials model:
    * [checkpoint](https://download.01.org/opencv/openvino_training_extensions/models/formula_recognition/polynomials_handwritten_0166.pth)
    * digits, letters, upper indices are supported
    * to use this model, please, change model path in the corresponding config file:
    ```
    model_path: <path to the model>
    ```
The model can be used for recognizing handwritten polynomial equations.
All the above models can be used for aftertuning or as ready for inference models. To provide maximum quality at recognizing formulas, it is highly recommended to preprocess image - simply binarize it:
```
val_transform_list:
    - name: TransformBin
      threshold: 100
```
You can find other prepocessing at [this file](im2latex/data/utils.py).
Sample images in the [data](../../data) section of this repo are already preprocessed, you can look at the examples.

#### Evaluation-specific parameters
- `split_file` - name of the file with labels (note: physical file name should end with `_filter.lst`). Default is `validate`

> - `target_metric` - target value of the metric. Used in tests. For test to pass, result value should be greater or equal than `target_metric`

Reviewer 2: not sure that it is needed in tests, it is needed for tests only.

Author: Removed from config file

- `train_transforms_list` - similiar to `val_transforms_list`
- `epochs` - number of epochs to train

One can use some pretrained models. Right now two models are available:
* medium model:
    * [checkpoint link](https://download.01.org/opencv/openvino_training_extensions/models/formula_recognition/medium_photograped_0185.pth)
    * digits, letters, some greek letters, fractions, trigonometric operations are supported; for more details, please, look at corresponding vocab file
    * to use this model, please, change these fields in the config file:
    ```
    backbone_config:
        arch: resnext50_32x4d
        disable_layer_3: true
        disable_layer_4: true
    model_path: <path to the model>
    vocab_path: vocabs/vocab_medium.json

>     ```

Reviewer 3: Why don't you just put the ready to use config file in the repo?..

Author: Updated config to use medium photographed model by default

> The model can be used either for recognizing rendered formulas and for recognizing scanned formulas (e.g. from a scanner or from a phone camera)

Reviewer 3: either ... or ...
both ... and ...

Author: Fixed

Reviewer 3: How the user's supposed to choose one of these two options?

Author: Reformulated this sentence, it was meant that model is able to recognize both rendered and scanned formulas, and user does not need to make any choice, model supports this functionality by default.

* handwritten polynomials model:
    * [checkpoint](https://download.01.org/opencv/openvino_training_extensions/models/formula_recognition/polynomials_handwritten_0166.pth)
    * digits, letters, upper indices are supported
    * to use this model, please, change these field in the config file:
    ```
    backbone_config:
        arch: resnext50_32x4d
        disable_layer_3: false
        disable_layer_4: true
    model_path: <path to the model>
    vocab_path: vocabs/vocab_handwritten_polynomials.json
    ```
The model can be used for recognizing handwritten polynomial equations.

> All the above models can be used for aftertuning or as ready for inference models. To provide maximum quality at recognizing formulas, it is highly recommended to preprocess image - simply binarize it, you can find corresponding prepocessing at [this file](im2latex/data/utils.py). Just state the desired preprocessing at the corresponding section of the config file (train, eval, etc).

Reviewer 3: Once again, my understanding is that config file + checkpoint should be enough. No manual tweaking.

Author: For medium model it is true, but polynomials model differs from medium one not only in weights, but in architecture too. So, we have to change the configuration.

Reviewer 3: Ok, take your time and think about it.

Author: What is your proposal? Create an additional config file for handwritten polynomials model?

Author: Added config for handwritten polynomials
